File: Topic10_Queque/Main.py
class Queue:
    def __init__(self, capacity):
        self.queue = [None] * capacity
        self.front = 0
        self.rear = -1
        self.size = 0
        self.capacity = capacity

    def enqueue(self, data):
        if self.is_full():
            raise QueueFullError("Queue is full!")
        self.rear = (self.rear + 1) % self.capacity
        self.queue[self.rear] = data
        self.size += 1

    def dequeue(self):
        if self.is_empty():
            raise QueueEmptyError("Queue is empty!")
        data = self.queue[self.front]
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return data

    def peek(self):
        if self.is_empty():
            raise QueueEmptyError("Queue is empty!")
        return self.queue[self.front]

    def is_empty(self):
        return self.size == 0

    def is_full(self):
        return self.size == self.capacity

    def print_queue(self):
        if self.is_empty():
            print("Queue is empty")
        else:
            print("Queue elements:", *self.queue[self.front:(self.front + self.size) % self.capacity])


class QueueFullError(Exception):
    pass

class QueueEmptyError(Exception):
    pass

if __name__ == "__main__":
    queue = Queue(5)

    queue.enqueue(10)
    queue.enqueue(20)
    queue.enqueue(30)
    queue.enqueue(40)
    queue.enqueue(50)

    print("Peek:", queue.peek())

    queue.dequeue()
    queue.dequeue()

    queue.enqueue(60)
    queue.enqueue(70)

    queue.print_queue()

File: Topic10_Queque/Main.java
package Topic10_Queue;

class Queue {
    private int rear;
    private int front;
    private int size;
    private final int capacity;
    private final int[] queue;

    public Queue() {
        rear = -1;
        front = 0;
        size = 0;
        capacity = 4;
        queue = new int[capacity];
    }

    public String enqueue(int data) {
        if (isFull()) {
            return "Queue Overflow: Cannot enqueue " + data;
        } else {
            rear = (rear + 1) % capacity;
            queue[rear] = data;
            size++;
            return "Enqueued: " + data;
        }
    }

    public String dequeue() {
        if (isEmpty()) {
            return "Queue Underflow: Cannot dequeue";
        } else {
            int data = queue[front];
            front = (front + 1) % capacity;
            size--;
            return "Dequeued: " + data;
        }
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == capacity;
    }

    public int peek() {
        if (isEmpty()) {
            System.out.println("Queue is empty.. cannot access peek value");
            return -1;
        } else {
            return queue[rear];
        }
    }

    public void printValues() {
        if (isEmpty()) {
            System.out.println("Queue is empty");
        } else {
            System.out.print("Queue elements: ");
            for (int i = 0; i < size; i++) {
                System.out.print(queue[(front + i) % capacity] + " ");
            }
            System.out.println();
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Queue queue = new Queue();

        System.out.println(queue.dequeue());

        System.out.println(queue.enqueue(10));
        System.out.println(queue.enqueue(20));
        System.out.println(queue.enqueue(30));
        System.out.println(queue.enqueue(40));
        queue.printValues();

        System.out.println(queue.dequeue());
        System.out.println(queue.dequeue());
        queue.printValues();

        System.out.println(queue.enqueue(50));
        System.out.println(queue.enqueue(60));
        queue.printValues();

        System.out.println(queue.enqueue(70));
        queue.printValues();
    }
}
